### DISCUSSION TITLE
casdoor - casbin auth portal

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#3449] [Show and tell]
@ganigeorgiev During your break from working on Pocketbase full-time, please check out casbin and casdoor.  It could be a great way of adding the collection expansion notation to determine group membership to routes both built-in and custom.  It can be used with or without normal API filters.

https://github.com/casdoor/casdoor

To be clear, I'm using https://github.com/casbin/casbin directly in Pocketbase with embedded csv files.  I only referred to casdoor because it could provide hints about how to integrate it into a Pocketbase Admin UI

**Comment by dschissler** [comment]
@ganigeorgiev 

This is part of my policy file.  I'm using a form of rule that allows for route parameters that seem to be fully compatible with echo.

It's all very easy to use.  The tricky part is based around what is considered a group, and how that is stored, and what happens if a route is blocked.  I've created two differently middlewares that allow me to redirect on failure, and another that just hard blocks.

I think that a hell of a lot could be done with this by using something like the Filter API syntax to expand into group membership.

```csv
p, /appi/site/:site/members, @normal, GET

# admin controller
p, /admin, @admin, GET
p, /admin/*, @admin, GET
p, /admin/api/*, @admin, (GET|POST)

######### GROUPS #########
g, @unverified, @anon
g, @normal, @unverified
g, @power, @normal
g, @admin, @power
g, @super, @admin

# root can do everything
p, /*, @root, (GET|POST|PUT|PATCH|DELETE|OPTIONS)
```

So the model looks like this.  It's a completely impossible prospect until you really spend the hours to understand what it's doing.  It's great because you can mix and match styles.  In this case I think that `keyMatch2` is what gives me the echo-like routes.

```
[request_definition]
r = sub, obj, act

[policy_definition]
p = obj, sub, act

[role_definition]
g = _, _

[policy_effect]
e = some(where (p.eft == allow))

[matchers]
m = g(r.sub, p.sub) && keyMatch2(r.obj, p.obj) && regexMatch(r.act, p.act)
```

So I think that something like this in Pocketbase (for non-admin UI routes) would really take Pocketbase to the next level.  It's also fully compatible with the amazing JS server work that you did.

It could also be used to tighten things up for normal user collection API.

**Reply by YugoCode** [reply]
I see it supports LDAP, would this for example also allow authentication through Active Directory in a corporate environment?

**Reply by dschissler** [reply]
@YugoCode https://casbin.org/docs/role-managers/

Yes.  There are all sorts of adapters.  I wrote my own adapter to read from the GO-embedded .csv and .conf files.
